packages/text-field/src/TextField/TextField.tsx

```diff
@@ -0,0 +1,135 @@
+import { useComposedRefs } from "@telegraph/compose-refs";
+import { Icon } from "@telegraph/icon";
+import { clsx } from "clsx";
+import React from "react";
+
+import { COLOR, SIZE } from "./TextField.constants";
+
+type TextFieldProps = Omit<
+  React.HTMLAttributes<HTMLInputElement>,
+  "value" | "onChange"
+> & {
+  size?: "1" | "2" | "3";
+  variant?: "outline" | "ghost";
+  disabled?: boolean;
+  error?: boolean | string;
+  value?: string;
+  onChange?: (
+    value: string,
+    { event }: { event: React.ChangeEvent<HTMLInputElement> },
+  ) => void;
+  leadingIcon?: React.ComponentProps<typeof Icon>;
+  TrailingAction?: React.ReactElement;
+};
+
+type TextFieldRef = HTMLInputElement;
+
+type InternalProps = {
+  id: string;
+  state: "default" | "error" | "disabled";
+  size: TextFieldProps["size"];
+  variant: TextFieldProps["variant"];
+  message?: string;
+};
+
+const TextField = React.forwardRef<TextFieldRef, TextFieldProps>(
+  (
+    {
+      size = "2",
+      variant = "outline",
+      value,
+      onChange,
+      leadingIcon,
+      TrailingAction,
+      className,
+      disabled,
+      error,
+      ...props
+    },
+    forwardedRef,
+  ) => {
+    const inputRef = React.useRef<HTMLInputElement>(null);
+    const composedRefs = useComposedRefs(inputRef, forwardedRef);
+
+    const [internalState, setInternalState] =
+      React.useState<InternalProps["state"]>("default");
+
+    React.useEffect(() => {
+      if (disabled) {
+        setInternalState("disabled");
+      } else if (error) {
+        setInternalState("error");
+      } else {
+        setInternalState("default");
+      }
+    }, [disabled, error]);
+
+    // Adds default props to the component, which can be overridden by the caller
+    const TrailingActionComponent = TrailingAction
+      ? React.cloneElement(TrailingAction, {
+          size,
+          color: "gray",
+          variant: "ghost",
+          ...TrailingAction?.props,
+        })
+      : null;
```

Comment:
This takes in the `TrailingAction` that is passed in and applies some default props to it.